### PR TITLE
MINOR: make flush no-op as we don't need to call flush on commit.

### PR DESCRIPTION
**Pull request by bbejeck**
In the event of a crash, we always restore the data from the backing changelog.  So it seems that we don't need to 
 persist all data to disk by calling flush when committing.  Frequent flushing leads to a large number of small files for compaction increasing
 compaction pressure.   This PR will perform benchmarks to see if there is any performance gain in not calling `flush()` each time we commit.

### CODE REVIEW

**Comment by bbejeck** [comment]
@guozhangwang @dguy  changes for not flushing during commit calls.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 1]
```diff
@@ -356,11 +356,7 @@ private boolean isOverflowing(long value) {
 
     @Override
     public synchronized void flush() {
-        if (db == null) {
-            return;
-        }
-        // flush RocksDB
-        flushInternal();
+        //NO-OP method since we don't flush on commits
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: nit: space after `//` in comments and we do not need to capitalize letters.

**Review by guozhangwang — COMMENTED** [review]
@bbejeck could you conduct some perf benchmarks (e.g. the `SimpleBenchmark` with shorter commit intervals) to illustrate how effective is this optimization?

**Review by dguy — COMMENTED** [review]
I guess we don't need to flush the store on close? Overall LGTM

**Comment by guozhangwang** [comment]
Upon close `flush` will be called anyways internally I think, but worth double check the RocksDB source doc.

**Comment by bbejeck** [comment]
TLDR; surprisingly it looks like close doesn't call flush,  adding `flushInternal` call to `RocksDBStore# close()`.

close sequence:
      `RocksDB.close()` -> 
     `disposeInternal()`  ->
      `disposeInteral(long nativeHandle)` 

which ultimately makes its way to  `rocksjni.cc` line 1492 `Java_org_rocksdb_RocksDB_disposeInternal` and  deletes the underlying object pointer.

EDIT: Nevermind, looks like the destructor is called when deleting underlying object pointer, but I think we should still leave in the flush call on close.

**Comment by guozhangwang** [comment]
> TLDR; surprisingly it looks like close doesn't call flush, adding flushInternal call to RocksDBStore# close().

Hmm.. this is surprising, as what it claims differently: https://github.com/facebook/rocksdb/blob/master/include/rocksdb/options.h#L870

**Comment by bbejeck** [comment]
I came to the same conclusion, looking at different parts of the docs.  I'll remove the flush call on close.

**Comment by guozhangwang** [comment]
I think `delete DB` itself actually does the flush:

https://github.com/facebook/rocksdb/blob/master/db/db_impl.cc#L260

Note it implements the `~DB()` function used for deallocate the pointer.

**Comment by bbejeck** [comment]
Here's a link to a spread sheet showing result from _**local**_ performance tests on my laptop using `SimpleBenchmark` NOTE this should not be used for comparison to other frameworks or used for performance characteristics of KS itself.  
https://docs.google.com/spreadsheets/d/1DURE5-9rVQtD_CVPj5dtrfP2Qu3Tx05ZSQxu_hHcm7Y/edit?usp=sharing

**Comment by guozhangwang** [comment]
@bbejeck Thanks for the shared benchmarks. They were really helpful.

One thing that I am not clear is that with 10 sec commit interval, flushing is actually significantly better than no-flushing (i.e. G43/44 and J43/44); I would expect that they do not have obvious difference in this case, and I'm wondering if it is just the variance of the benchmark itself? E.g. how many runs have you done for each comparison, and did the experiment have enough data to sustain at least 10 seconds (I remember the SimpleBenchmark only 10M records in total).

Once that is cleared I think this patch is good to go.

**Comment by enothereska** [comment]
@bbejeck could we add a top level description to this PR? I don't immediately understand why flushing is not needed. This is conceptually a really big change, I just want to make sure I (and others) understand it. I think you've explained it in the past, but I've forgotten :(

**Comment by bbejeck** [comment]
@enothereska - added a top-level comment

**Comment by bbejeck** [comment]
@guozhangwang - re-running 10 sec benchmarks with more data

**Comment by enothereska** [comment]
@bbejeck interesting thanks. If I understand it right, upon commit, we still store the checkpoint file that indicates how much data we're got in RocksDb and where we need to start off (what offset) after a crash. Now, if we don't flush RocksDb and we crash, will it be the case that the checkpoint file will have information that is not reflected in RocksDB? Thanks.

**Comment by bbejeck** [comment]
@enothereska updated to only skip flush when EOS is enabled.  @guozhangwang re-ran benchmarks with more data for 10sec commits, will updated spreadsheet.

**Review by enothereska — APPROVED** [review]

**Comment by bbejeck** [comment]
@guozhangwang I'd like to rebase this PR and run one more time for clarification.

**Comment by bbejeck** [comment]
@guozhangwang 
re-ran the tests and for the 10 Sec commit phase (20 million records you'll see the latency numbers are better for 10 commit phases) and the numbers seem to align a little better.  However the results on  G43/44 and J43/44 still seem strange and I think these tests should be run separately vs part of the entire test suite.  

I have some ideas on why this could happen, but since the number look good for lower commit intervals (which is what we use of EOS) I think we can merge while look into this some more.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5368/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5353/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by guozhangwang** [comment]
Thanks @bbejeck . Looks like G43/44 and J43/44 are outliers since they do not linearly regress with other data points. I will merge the PR to trunk as is as we can continue working on the checkpointing / flushing optimizations.
